### PR TITLE
UnnecessaryAny - Add rule to prevent usage of `any { }` when simple `contains` can be …

### PR DESCRIPTION
**Pull request by atulgpt**
Fixes https://github.com/detekt/detekt/issues/6106

### CODE REVIEW

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAnySpec.kt [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+    fun `does not report any with safe call which is used for checking presence of a element`() {
+        val code = """
+            fun test(list: List<Int>, value: Int?) {
+                list.any { value?.equals(it) == true }
+            }
+        """.trimIndent()
+        val actual = subject.compileAndLintWithContext(env, code)
+        assertThat(actual).isEmpty()
+    }
+
```
Comment (BraisGabin) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAnySpec.kt:131]: I think that we could remove this test. I think that this code should be flagged but it seems so edge case to me that probably it doesn't need the effort. But we at least should don't have a test that say "this shouldn't be flagged".

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAnySpec.kt [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+        val code = """
+            fun test(list: List<Int>, value: Int) {
+                list.any { value == 2 }
+                list.any { 2 == value }
+            }
+        """.trimIndent()
+        val actual = subject.compileAndLintWithContext(env, code)
+        assertThat(actual).isEmpty()
+    }
+
```
Comment (BraisGabin) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAnySpec.kt:154]: I'm not 100% sure how but this is clearly an "UnnecessaryAny". If the lambda is not related with `it` you shouldn't use `any` for sure.

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAnySpec.kt (hunk 1 above)
Comment (atulgpt) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAnySpec.kt:131]: Actually, this should not be reported by the rule. I made below example where usage of `any` and `contain` prints different thing

```kotlin
fun main() {
    val list = listOf(1,2,3, null)
    
    println("${test1(list, null)}, ${test2(list, null)}")
    println("${test1(list, 1)}, ${test2(list, 1)}")
}

fun test1(list: List<Int?>, value: Int?): Boolean {
    return list.any { value?.equals(it) == true }
}

fun test2(list: List<Int?>, value: Int?): Boolean {
    return list.contains(value)
}
```

Here list is of nullable type and the list contains null. So let me instead update the TC and make the `Int` to `Int?`

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAnySpec.kt (hunk 2 above)
Comment (atulgpt) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAnySpec.kt]: I have updated the code to detect when it is not used in the `any`

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAnySpec.kt [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+    @Test
+    fun `does report any when it is not used`() {
+        val code = """
+            fun test(list: List<Int>, value: Int) {
+                list.any { value == 2 }
+                list.any { 2 == value }
```
Comment (BraisGabin) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAnySpec.kt:148]: We should change the error message in this type of errors, don't you think?

I don't know what we should recommend here. Probably just say that `any` shouldn't be used like this?

Comment (atulgpt) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAnySpec.kt:173]: Hi, I have updated the msg as per msg used in the `UnnecessaryLet`. Updated msg is `any {  }` expression can be omitted`.

Also, fix some bugs which I found while doing this change.

**Review by BraisGabin — APPROVED** [review]
LGTM! Thanks!

**Comment by cortinico** [comment]
@BraisGabin I also won't be able to pick this one as it's a new rule
